### PR TITLE
Fix: Memory leakage in `QkCountOps` creation.

### PR DESCRIPTION
**Pull request by raynelfss**
### Summary
The following commits attempt to fix some memory issues that the struct `QkOpCounts`. The `qk_opcounts_free` method would segfault if an empty object is passed. The function now checks whether the passed option includes a null pointer or an empty instance. 

The commits also change how the `data` attribute is generated without using `std::mem::forget`, but instead collecting the `OpCount` vec into a boxed_slice and consuming it into a `Box`. Inspired by what was done for #14884.

### Details and comments

Found during development of #14710.

### CODE REVIEW

File: test/c/test_circuit.c [hunk 1]
```diff
@@ -27,6 +27,15 @@ int test_empty(void) {
     uint32_t num_qubits = qk_circuit_num_qubits(qc);
     uint32_t num_clbits = qk_circuit_num_clbits(qc);
     size_t num_instructions = qk_circuit_num_instructions(qc);
+
+    // This snippet tests that an empty ops_count is possible
+    QkOpCounts counts = qk_circuit_count_ops(qc);
+    if (counts.len != 0) {
+        printf("The number of instructions %zu is not 0", num_instructions);
+        return EqualityError;
```
Comment (Cryoris) [inline test/c/test_circuit.c]: Hehe the PR fixing a memory leak introduces one in the tests 😛 This needs to free `counts` and `qc` 🙂

File: crates/cext/src/circuit.rs [hunk 2]
```diff
@@ -782,16 +782,19 @@ pub unsafe extern "C" fn qk_circuit_count_ops(circuit: *const CircuitData) -> Op
...
-        .map(|(name, count)| OpCount {
-            name: CString::new(name).unwrap().into_raw(),
-            count,
-        })
-        .collect();
+    let mut output = {
+        let vec: Vec<OpCount> = count_ops
+            .into_iter()
+            .map(|(name, count)| OpCount {
+                name: CString::new(name).unwrap().into_raw(),
```
Comment (Cryoris) [inline crates/cext/src/circuit.rs:795]: This allocates a CString, shouldn't we also free this in the process? I.e. first load the OpCount objects, then iterate and free the strings, and then free the vector itself?

Comment (raynelfss) [reply crates/cext/src/circuit.rs:795]: FYI, @Cryoris is adding this in a separate [PR](https://github.com/raynelfss/qiskit/pull/7)

File: test/c/test_circuit.c (hunk 1 above)
Comment (raynelfss) [reply test/c/test_circuit.c]: Ironic, addressed in 0d941edf1397e2b5065050e69cab2365553866a8

File: crates/cext/src/circuit.rs [hunk 3]
```diff
@@ -782,16 +782,19 @@ pub unsafe extern "C" fn qk_circuit_count_ops(circuit: *const CircuitData) -> Op
...
+                count,
+            })
+            .collect();
+        vec.into_boxed_slice()
```
Comment (eliarbel) [inline crates/cext/src/circuit.rs:799]: I don't think we need this since we shouldn't care about the length field of `vec` itself, hence we shouldn't bother moving it to the heap. The length information is stored in `OpCounts::len` anyway. We only care about `vec`'s buffer which is taken using `as_mut_ptr` below.

File: crates/cext/src/circuit.rs [hunk 4]
```diff
@@ -782,16 +782,19 @@ pub unsafe extern "C" fn qk_circuit_count_ops(circuit: *const CircuitData) -> Op
...
+    };
     let data = output.as_mut_ptr();
     let len = output.len();
-    std::mem::forget(output);
+    let _ = Box::into_raw(output);
```
Comment (eliarbel) [inline crates/cext/src/circuit.rs]: I think the original `std::mem::forget(output);` call was good, as it keeps the vec's buffer alive.

File: crates/cext/src/circuit.rs [hunk 5]
```diff
@@ -1008,14 +1011,29 @@ pub unsafe extern "C" fn qk_circuit_instruction_clear(inst: *mut CInstruction) {
...
-        let _ = Box::from_raw(data);
+pub unsafe extern "C" fn qk_opcounts_clear(op_counts: *mut OpCounts) {
+    // SAFETY: The user guarantees the input is a valid OpCounts pointer.
+    let op_counts = unsafe { mut_ptr_as_ref(op_counts) };
+
+    if op_counts.len > 0 && !op_counts.data.is_null() {
+        // SAFETY: We load the box from a slice pointer created from
+        // the raw parts from the OpCounts::data attribute.
+        unsafe {
+            let vec = Box::from_raw(std::slice::from_raw_parts_mut(
```
Comment (eliarbel) [inline crates/cext/src/circuit.rs]: How about something like this instead:

```rust
            let vec = Vec::from_raw_parts(
                op_counts.data,
                op_counts.len,
                op_counts.len
            );
```

File: crates/cext/src/circuit.rs [hunk 6]
```diff
@@ -782,16 +782,19 @@ pub unsafe extern "C" fn qk_circuit_count_ops(circuit: *const CircuitData) -> Op
     // SAFETY: Per documentation, the pointer is non-null and aligned.
```
Comment (eliarbel) [inline crates/cext/src/circuit.rs:788]: I would add a note to the doc comment in the nature of: 

```rust
/// Note:  ``qk_opcounts_clear`` should be called with the returned ``QkOpsCounts`` struct
/// to clear any heap-allocated data generated by this function.
```

File: crates/cext/src/circuit.rs [hunk 7]
```diff
@@ -1008,14 +1011,29 @@ pub unsafe extern "C" fn qk_circuit_instruction_clear(inst: *mut CInstruction) {
...
+            // free the allocated strings in each OpCount
+            for count in vec.iter() {
+                if !count.name.is_null() {
+                    let _ = CString::from_raw(count.name as *mut c_char);
+                }
+            }
+            // the variable vec goes out of bounds and is freed too
+        }
     }
+    op_counts.len = 0;
```
Comment (eliarbel) [inline crates/cext/src/circuit.rs:1040]: I wonder if we really need this... I would think it's reasonable to assume that after calling `qk_opcounts_clear` the `QkOpCounts` struct is no longer meant to be used, so why should we bother resetting its fields?

File: test/c/test_circuit.c [hunk 8]
```diff
@@ -254,6 +263,40 @@ int test_gate_num_params(void) {
...
+    double params[1] = {0.12};
+    qk_circuit_gate(qc, QkGate_CRX, qubits, params);
+    QkOpCounts c2 = qk_circuit_count_ops(qc);
+
+    if (c2.len != 1) {
+        result = EqualityError;
+    }
+    qk_opcounts_clear(&c2);
+
+    // check that after clearing, the object is still valid
```
Comment (eliarbel) [inline test/c/test_circuit.c:295]: See comment above. I think this guarantee is an overkill.

**Review by eliarbel — COMMENTED** [review]
Thank @raynelfss for handling this. Unless I'm missing something, it seems to me that `qk_circuit_count_ops` is already OK, and we only need to update the `free/clear` function. If so see my line comments.

File: crates/cext/src/circuit.rs (hunk 7 above)
Comment (Cryoris) [reply crates/cext/src/circuit.rs:1040]: What concerned me here a bit is that the `QkOpCounts` is still valid to use after calling `qk_opcounts_clear` but in an inconsistent state -- the length is still the old value but the memory is cleared and not readable. I rather didn't want to make any assumptions about whether the user means to discard the object or not 😄

Comment (eliarbel) [reply crates/cext/src/circuit.rs:1040]: Yes, makes sense, though this holds for any other local pointer the user has after calling the relevant `*_free` function on it. I mean, if the user really wants to live on the edge, the C API gives them a lot of opportunities for this 😉 
Anyway, it's not critical, just seemed an overkill to me. Alternatively, we can add in the `qk_opcounts_clear` function doc comment that the struct is not valid for use after clearing the data.

Comment (Cryoris) [reply crates/cext/src/circuit.rs:1040]: I agree for `*_free`, this object is not safe to use anymore after -- but I'd say `free` implies that whereas `clear` to me means that it's just sort of reset, not unsafe to use. A bit related to Jake's posted message on https://github.com/git/git/blob/c44beea485f0f2feaf460e2ac87fdd5608d63cf0/Documentation/CodingGuidelines#L621-L637 😄

Comment (eliarbel) [reply crates/cext/src/circuit.rs:1040]: Ok, let's leave the clearing calls and test in place.

File: test/c/test_circuit.c (hunk 8 above)
Comment (eliarbel) [reply test/c/test_circuit.c:295]: In light of this: https://github.com/Qiskit/qiskit/pull/14930#discussion_r2297680631, let's keep it.

File: crates/cext/src/circuit.rs (hunk 3 above)
Comment (Cryoris) [reply crates/cext/src/circuit.rs:799]: As I understand it, the data of a `Vec` is allocated on the heap anyways, no? Then the main difference between the two approaches here is how much memory is actually allocated and how we need to free it later:
```rust
// Vec<OpCount> allocates according to capacity on the heap 
// (which can be larger than it's len)
let vec = ...;  
let data = vec.as_mut_ptr();
let len = vec.len();
let cap = vec.capacity();  // we don't store this right now, so we can't correctly de-allocate, can we?
std::mem::forget(vec);  // forget about all the memory: of size capacity

// for freeing we need to know the capacity the vector had -- which we're not doing right now
let _ = Vec::from_raw_parts(data, len, cap);
```
vs.
```rust
// Box<[OpCount]> allocates only ``len`` elements on the heap
let slice: Box<[OpCount]> = vec.into_boxed_slice();
let data = slice.as_mut_ptr();
let len = slice.len();
let _ = Box::into_raw(slice);  // forget about all the memory: of size len

// we can properly free knowing only the len
let _ = Box::from_raw(::std::slice::from_raw_parts(data, len));
```

In https://github.com/Qiskit/qiskit/pull/14884 I think we generically moved to exposing `Box<[T]>` to C due to this (which tbh I didn't fully understand until trying to understand your comment above and going down this rabbit hole 🙂)

Comment (eliarbel) [reply crates/cext/src/circuit.rs:799]: This ping pong also helps me understand this better 😄. You're right about the `capacity` info, we don't store it but need it later for deallocating the data when using `Vec::from_raw_parts`. So, calling `vec.shrink_to_fit()` here should do the trick, and we can use `vec.len()` as the capacity size when calling `from_raw_parts`. This is likely an O(n) operation for reallocation + copying. 

`into_boxed_slice` indeed boxes the buffer such that at the end length == "capacity" but for doing so it calls `shrink_to_fit` [under the hood](https://github.com/rust-lang/rust/blob/4356e83c77cb28113411fb8e2219127f708095b5/library/alloc/src/vec/mod.rs#L1567) anyway, so complexity-wise, these are equivalent. 

I guess at the end it's more about style preference
@raynelfss

File: crates/cext/src/circuit.rs (hunk 5 above)
Comment (Cryoris) [reply crates/cext/src/circuit.rs]: Loading this as `Box<[T]>` again feels safer than loading this as `Vec`, I could rename the variable to make it clearer that it's not a vec in fact? 🙂 

Edit: done in c81ac55

File: crates/cext/src/circuit.rs (hunk 3 above)
Comment (Cryoris) [reply crates/cext/src/circuit.rs:799]: I'd prefer keeping the `into_boxed_slice`, if just to keep a consistent way of handling vectors. It also seems a bit safer to code since we can't forget about e.g calling `shrink_to_fit` 😄

File: crates/cext/src/circuit.rs (hunk 4 above)
Comment (Cryoris) [reply crates/cext/src/circuit.rs]: I'll resolve this since it's connected to the above discussion 🙂

File: crates/cext/src/circuit.rs (hunk 6 above)
Comment (Cryoris) [reply crates/cext/src/circuit.rs:788]: Added in 3d6e87f

**Comment by raynelfss** [comment]
@Cryoris will be taking over from me on this effort based on internal discussion.

File: releasenotes/notes/c-api-ops-count-bug-f633d349de315e89.yaml [hunk 9]
```diff
@@ -0,0 +1,6 @@
+---
+fixes:
+  - |
+    Fix memory leakage issues during the creation of
```
Comment (eliarbel) [inline releasenotes/notes/c-api-ops-count-bug-f633d349de315e89.yaml]:
```suggestion
    Fixed memory leakage issues during the creation of
```

File: releasenotes/notes/c-api-ops-count-bug-f633d349de315e89.yaml [hunk 10]
```diff
@@ -0,0 +1,6 @@
...
+    :cpp:struct:`QkOpCounts` instances and during any calls to
```
Comment (eliarbel) [inline releasenotes/notes/c-api-ops-count-bug-f633d349de315e89.yaml]:
```suggestion
    a :cpp:struct:`QkOpCounts` instance and during any calls to
```

File: crates/cext/src/circuit.rs [hunk 11]
```diff
@@ -761,6 +761,10 @@ pub unsafe extern "C" fn qk_circuit_unitary(
 /// @ingroup QkCircuit
 /// Return a list of string names for instructions in a circuit and their counts.
 ///
+/// To properly free the memory allocated by the struct, you should call ``qk_opcounts_clear``.
+/// Dropping the ``QkOpCounts`` struct without doing so will leave the stored array of ``OpCount``
+/// allocated and produce a memory leak.
+///
 /// @param circuit A pointer to the circuit to get the counts for.
 ///
 /// @return An ``OpCounts`` struct containing the circuit operation counts.
```
Comment (eliarbel) [inline crates/cext/src/circuit.rs]:
```suggestion
/// @return An ``QkOpCounts`` struct containing the circuit operation counts.
```

File: crates/cext/src/circuit.rs [hunk 12]
```diff
@@ -761,6 +761,10 @@ pub unsafe extern "C" fn qk_circuit_unitary(
 /// @ingroup QkCircuit
 /// Return a list of string names for instructions in a circuit and their counts.
 ///
+/// To properly free the memory allocated by the struct, you should call ``qk_opcounts_clear``.
+/// Dropping the ``QkOpCounts`` struct without doing so will leave the stored array of ``OpCount``
```
Comment (eliarbel) [inline crates/cext/src/circuit.rs]:
```suggestion
/// Dropping the ``QkOpCounts`` struct without doing so will leave the stored array of ``QkOpCount``
```

File: test/c/test_circuit.c [hunk 13]
```diff
@@ -254,6 +263,40 @@ int test_gate_num_params(void) {
...
+ * Test edge cases for getting the op counts.
+ */
+int test_get_gate_counts(void) {
+    QkCircuit *qc = qk_circuit_new(3, 3);
+
+    // test empty circuit
+    int result = Ok;
+    QkOpCounts c1 = qk_circuit_count_ops(qc);
+    if (c1.len != 0) {
+        result = EqualityError;
```
Comment (eliarbel) [inline test/c/test_circuit.c:276]: Why not failing fast and "goto"ing to a cleanup label?

Comment (Cryoris) [reply test/c/test_circuit.c:276]: `goto` statements are banned in my vocabulary, I need to remember that sometimes they are ok 😉 added!

File: crates/cext/src/circuit.rs (hunk 3 above)
Comment (eliarbel) [reply crates/cext/src/circuit.rs:799]: After consulting with `Miri` and @jakelishman, for the `into_boxed_sliced` approach this fragment should make everyone happy:
```rust
    let output = {
        let vec: Vec<OpCount> = count_ops
            .into_iter()
            .map(|(name, count)| OpCount {
                name: CString::new(name).unwrap().into_raw(),
                count,
            })
            .collect();
        vec.into_boxed_slice()
    };
    let len = output.len();
    let raw = Box::into_raw(output) as *mut OpCount;
    let data = raw;
    OpCounts { data, len }
```

Comment (Cryoris) [reply crates/cext/src/circuit.rs:799]: Nice! Done in c81bcc2.

**Review by eliarbel — DISMISSED** [review]
LGTM, thanks!

**Review by eliarbel — APPROVED** [review]
